rules/legacy/src/Rector/Include_/AddTopIncludeRector.php

```diff
@@ -0,0 +1,200 @@
+use PhpParser\Node\Stmt\GroupUse;
+use PhpParser\Node\Stmt\InlineHTML;
+use PhpParser\Node\Stmt\Namespace_;
+use PhpParser\Node\Stmt\Nop;
+use PhpParser\Node\Stmt\Use_;
+use PhpParser\Node\Stmt\UseUse;
+use Rector\Core\Exception\Rector\InvalidRectorConfigurationException;
+use Rector\Core\Rector\AbstractRector;
+use Rector\Core\RectorDefinition\CodeSample;
+use Rector\Core\RectorDefinition\RectorDefinition;
+use Rector\NodeTypeResolver\Node\AttributeKey;
+use ReflectionClass;
+
+/**
+ * @see https://github.com/rectorphp/rector/issues/3679
+ *
+ * @see \Rector\Legacy\Tests\Rector\Include_\AddTopIncludeRector\AddTopIncludeRectorTest
+ */
+final class AddTopIncludeRector extends AbstractRector
+{
+    /**
+     * @var bool
+     */
+    private $completed = false;
+
+    /**
+     * @var string
+     */
+    private $file = '';
+
+    /**
+     * @var string
+     */
+    private $lastFileProcessed = '';
+
+    /**
+     * @var int
+     */
+    private $type = 0;
+
+    /**
+     * @var string[]
+     */
+    private $classNodes = [
+        Declare_::class => true,
+        DeclareDeclare::class => true,
+        GroupUse::class => true,
+        InlineHTML::class => true,
+        Namespace_::class => true,
+        Use_::class => true,
+        UseUse::class => true,
+    ];
+
+    /**
+     * @var string[] to match
+     */
+    private $patterns = [];
+
+    /**
+     * @var array
+     */
+    private $configuration = [];
+
+    public function __construct(array $configuration = [])
+    {
+        $this->configuration = $configuration;
+    }
+
+    public function getDefinition(): RectorDefinition
+    {
+        return new RectorDefinition('Adds an include file at the top of matching files but not class definitions', [
+            new CodeSample(
+                '<?php
+
+if (isset($_POST["csrf"])) {
+    processPost($_POST);
+}',
+                '<?php
+
+include "autoloader.php";
+
+if (isset($_POST["csrf"])) {
+    processPost($_POST);
+}',
+                [
+                    '$configuration' => [
+                        'type' => 'TYPE_INCLUDE_FILE',
+                        'file' => 'autoload.php',
+                        'match' => ['filePathWildCards'],
+                    ],
+                ]
+            ),
+        ]);
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        $this->initialize();
+
+        return [Node::class];
+    }
+
+    public function refactor(Node $node): ?Node
+    {
+        $fileInfo = $node->getAttribute(AttributeKey::FILE_INFO);
+        $path = $fileInfo ? $fileInfo->getRelativeFilePath() : '';
+
+        if ($path !== $this->lastFileProcessed) {
+            $this->lastFileProcessed = $path;
+            $this->completed = false;
+        }
+
+        if (! $this->matchFile($path)) {
+            $this->completed = true;
+
+            return null;
+        }
+
+        $name = get_class($node);
```

Comment:
```suggestion
        $nodeClass = get_class($node);
```

Reply:
So you are saying if I move to AbstractFileSystemRector, I can look for class and include (of the same file name), and if neither is found, I can insert before the first node and that should work?

I'll will check to see if that would handle InlineHTML before the first executable statement.  That is the only exception that might be an issue.

Reply:
> I can look for class and include (of the same file name), and if neither is found, I can insert before the first node and that should work?

Partially. It should also work with other includes of different files, e.g.

```diff
+include __DIR__ . '/autoloader.php' # be sure the paths is absolute, to prevent magic with relative paths 

 include 'some_other_file.php;

 if (...) {
    include 'yet_another.php;
 }
```